pkg/lib/operatorstatus/monitor.go

```diff
@@ -0,0 +1,221 @@
+// to send update notifications to it.
+//
+// The name of the clusteroperator resource to monitor is specified in name.
+func NewMonitor(name string, log *logrus.Logger, opClient operatorclient.ClientInterface, configClient configv1client.ConfigV1Interface) (Monitor, Sender) {
+	logger := log.WithField("monitor", "clusteroperator")
+	discovery := opClient.KubernetesInterface().Discovery()
+	releaseVersion := os.Getenv("RELEASE_VERSION")
+	names := split(name)
+
+	logger.Infof("monitoring the following components %s", names)
+
+	monitor := &monitor{
+		logger:         logger,
+		reporter:       newReporter(releaseVersion),
+		writer:         NewWriter(discovery, configClient),
+		notificationCh: make(chan *NotificationContext, defaultNotificationChannelSize),
+		names:          names,
+	}
+
+	return monitor, monitor
+}
+
+// NotificationContext contains all necessary information related to a notification.
+type NotificationContext struct {
+	// Name of the clusteroperator resource associated with this CSV.
+	Name string
+
+	// Current is the CSV for which we have received notification.
+	Current *v1alpha1.ClusterServiceVersion
+
+	// If there is an upgrade going on, Final is set to the latest version of
+	// the CSV that is replacing current.
+	// For a chain like this, Current (v1) -> v2 -> v3 -> Final (v4)
+	// Final will be set to the CSV linked to v4.
+	// Otherwise this value is set to nil.
+	Final *v1alpha1.ClusterServiceVersion
+
+	// Deleted indicates whether Current has been deleted.
+	Deleted bool
+}
+
+// Monitor is an interface that wraps the Run method.
+//
+// Run runs forever, it reads from an underlying notification channel and
+// updates an clusteroperator resource accordingly.
+// If the specified stop channel is closed the loop must terminate gracefully.
+type Monitor interface {
+	Run(stopCh <-chan struct{})
+}
+
+// Sender is an interface that wraps the Send method.
+//
+// Send can be used to send notification(s) to the underlying monitor. Send is a
+// non-blocking operation.
+// If the underlying monitor is not ready to receive the notification will be lost.
+// If the notification context specified is nil then it is ignored.
+type Sender interface {
+	Send(context *NotificationContext)
+}
+
+func (n *NotificationContext) String() string {
+	replaces := "<nil>"
+	if n.Final != nil {
+		replaces = n.Final.GetName()
+	}
+
+	return fmt.Sprintf("name=%s csv=%s deleted=%s replaces=%s", n.Name, n.Current.GetName(), strconv.FormatBool(n.Deleted), replaces)
+}
+
+type monitor struct {
+	notificationCh chan *NotificationContext
+	writer         *Writer
+	logger         *logrus.Entry
+	reporter       *reporter
+	names          []string
+}
+
+func (m *monitor) Send(context *NotificationContext) {
+	if context == nil {
+		return
+	}
+
+	select {
+	case m.notificationCh <- context:
+	default:
+		m.logger.Warn("monitor not ready to receive")
+	}
+}
+
+func (m *monitor) Run(stopCh <-chan struct{}) {
```

Comment:
nit: Do we want to use [`context.Context`](https://golang.org/pkg/context/) here instead?